File: packages/core/src/WardxCore.js
import { MetricsRegistry } from './metrics/MetricsRegistry.js';
import { EventBuffer } from './buffers/EventBuffer.js';
import { LogBuffer } from './buffers/LogBuffer.js';
import { ConfigStore } from './config/ConfigStore.js';
import { ExperimentResolver } from './config/ExperimentResolver.js';
import { subjectHash } from './config/hash.js';
import { FrameBuilder } from './frame/FrameBuilder.js';
import { InternalMetrics } from './internal/InternalMetrics.js';
import { NOOP_COUNTER } from './metrics/Counter.js';
import { NOOP_GAUGE } from './metrics/Gauge.js';
import { NOOP_HISTOGRAM } from './metrics/Histogram.js';
import { NOOP_DISTINCT } from './metrics/HyperLogLog.js';
import { startTimer } from './metrics/Timer.js';
import { emit } from './trace/emit.js';
import { wrapCounter, wrapDistinct, wrapGauge, wrapHistogram } from './trace/wrap.js';

export class WardxCore {
  constructor(settings) {
    this.settings = settings;
    this.stopped = false;
    this._tracer = settings.tracer ?? null;
    this._wrappers = this._tracer ? new WeakMap() : null;
    this.internal = new InternalMetrics();
    this.metrics = new MetricsRegistry({
      maxSeriesPerMetric: settings.maxSeriesPerMetric,
      maxDimensionKeys: settings.maxDimensionKeys,
      maxDimensionValueLength: settings.maxDimensionValueLength,
      defaultHistogramBuckets: settings.histogramBuckets,
      privacySalt: settings.privacySalt,
      onCardinalityDropped: () => {
        this.internal.cardinalityDropped += 1;
      }
    });
    this.events = new EventBuffer(settings.maxBufferedEvents);
    this.logs = new LogBuffer(settings.maxBufferedLogs);
    this.configStore = new ConfigStore();
    this.experiments = new ExperimentResolver({
      privacySalt: settings.privacySalt,
      stateMaxSubjects: settings.experimentStateMaxSubjects,
      onExposure: (payload) => {
        this.event('experiment.exposure', payload);
      }
    });
    this.seq = 0;
    this.pendingFrames = [];
    this.windowStart = Date.now();
    this._subjectId = null;
    this.log = {
      debug: (message, attrs) => this._log('debug', message, attrs),
      info: (message, attrs) => this._log('info', message, attrs),
      warn: (message, attrs) => this._log('warn', message, attrs),
      error: (message, attrs) => this._log('error', message, attrs)
    };
  }

  counter(name, dims) {
    return this._wrap(this.metrics.counter(name, dims), NOOP_COUNTER, (series, noop) =>
      wrapCounter(this._tracer, series, noop, name, dims)
    );
  }

  gauge(name, dims) {
    return this._wrap(this.metrics.gauge(name, dims), NOOP_GAUGE, (series, noop) =>
      wrapGauge(this._tracer, series, noop, name, dims)
    );
  }

  histogram(name, a, b) {
    return this._wrap(this.metrics.histogram(name, a, b), NOOP_HISTOGRAM, (series, noop) =>
      wrapHistogram(this._tracer, series, noop, name)
    );
  }

  distinct(name, dims) {
    return this._wrap(this.metrics.distinct(name, dims), NOOP_DISTINCT, (series, noop) =>
      wrapDistinct(this._tracer, series, noop, name, dims)
    );
  }

  timer(name, dims) {
    if (this._tracer === null) return this.metrics.timer(name, dims);
    return startTimer((duration, endDims) => {
      const merged = endDims ? { ...(dims || {}), ...endDims } : dims;
      this.histogram(name, merged).observe(duration);
    });
  }

  event(name, attrs) {
    const dropped = !this.events.push(name, attrs);
    if (dropped) this.internal.eventsDropped += 1;
    emit(this._tracer, 'event', { name, attrs: attrs ?? null, dropped });
  }

  _log(level, message, attrs) {
    const dropped = !this.logs.push(level, message, attrs);
    if (dropped) this.internal.logsDropped += 1;
    emit(this._tracer, 'log', { level, message, attrs: attrs ?? null, dropped });
  }

  _wrap(series, noopSentinel, factory) {
    if (this._tracer === null) return series;
    if (series === noopSentinel) return factory(series, true);
    let wrapped = this._wrappers.get(series);
    if (wrapped) return wrapped;
    wrapped = factory(series, false);
    this._wrappers.set(series, wrapped);
    return wrapped;
  }

  retentionActivity(userId) {
    if (typeof userId !== 'string' || userId.trim().length === 0) {
      throw new Error('retentionActivity requires a non-empty userId');
    }
    const salt = this.settings.privacySalt;
    if (typeof salt !== 'string' || salt.trim().length === 0) {
      throw new Error('retentionActivity requires a non-empty privacySalt');
    }
    this.event('retention.activity', {
      subject: subjectHash(salt, userId),
      salt: subjectHash(salt, 'wardx.retention.identity')
    });
  }

  identify(subjectId) {
    if (subjectId === undefined || subjectId === null) {
      this._subjectId = null;
      return;
    }
    if (typeof subjectId !== 'string' || subjectId.length === 0) {
      throw new Error('identify requires a non-empty subjectId');
    }
    this._subjectId = subjectId;
  }

  _subjectIdFrom(context) {
    if (context && context.subjectId !== undefined && context.subjectId !== null) {
      return context.subjectId;
    }
    return this._subjectId;
  }

  configGet(key, fallback, context) {
    if (!this.configStore.has(key)) return fallback;
    const remote = this.configStore.getRaw(key);
    const subjectId = this._subjectIdFrom(context);
    if (subjectId === undefined || subjectId === null) return remote;
    return this.experiments.resolve(
      key,
      remote,
      subjectId,
      this.configStore.experimentsByKey
    );
  }

  experimentGoal(name, context) {
    const subjectId = this._subjectIdFrom(context);
    if (subjectId === undefined || subjectId === null) {
      throw new Error('experiment.goal requires subjectId');
    }
    if (typeof name !== 'string' || name.length === 0) {
      throw new Error('experiment.goal requires a metric name');
    }
    const subject = this.experiments.hashSubject(subjectId);
    const assignment = this.experiments.exposedAssignmentForGoal(subjectId, name);
    if (assignment === null) return;
    const payload = {
      metric: name,
      subject,
      experiments: [{ experiment: assignment.experiment, variant: assignment.variant }]
    };
    if (context && context.value !== undefined) payload.value = context.value;
    this.event('experiment.goal', payload);
  }

  applyConfig(version, config) {
    this.configStore.applySnapshot({
      version,
      values: config.values,
      experiments: config.experiments
    });
    this.experiments.applySnapshot(this.configStore.experiments);
    this.internal.configVersion = version;
  }

  snapshotIfDirty() {
    if (
      !this.metrics.isDirty() &&
      this.events.length === 0 &&
      this.logs.length === 0 &&
      !this.internal.hasCounterActivity()
    ) {
      return null;
    }
    return this.snapshotFrame();
  }

  snapshotFrame() {
    const to = Date.now();
    const from = this.windowStart;
    this.windowStart = to;
    this.internal.eventsBuffered = this.events.length;
    this.internal.logsBuffered = this.logs.length;
    const metrics = this.metrics.snapshotAndReset();
    const events = this.events.swap();
    const logs = this.logs.swap();
    const internal = this.internal.snapshotAndReset();
    const frame = FrameBuilder.build({
      seq: this.seq + 1,
      from,
      to,
      metrics,
      events,
      logs,
      internal
    });
    const batch = FrameBuilder.splitToMaxBytes(frame, this.settings.maxFrameBytes);
    this.seq = batch.frames.at(-1).seq;
    this.pendingFrames.push(...batch.frames);
    for (let i = 0; i < batch.frames.length; i++) {
      const physical = batch.frames[i];
      emit(this._tracer, 'frame', {
        seq: physical.seq,
        from: physical.from,
        to: physical.to,
        counters: physical.metrics.counters.length,
        gauges: physical.metrics.gauges.length,
        histograms: physical.metrics.histograms.length,
        distincts: physical.metrics.distincts?.length || 0,
        events: physical.events.length,
        logs: physical.logs.length,
        droppedLogs: i === 0 ? batch.droppedLogs : 0,
        droppedEvents: i === 0 ? batch.droppedEvents : 0,
        droppedRows: i === 0 ? batch.droppedRows : 0
      });
    }
    return batch;
  }

  takePendingFrames() {
    const frames = this.pendingFrames;
    this.pendingFrames = [];
    return frames;
  }
}

File: clients/csharp/Runtime/Core/WardxCore.cs
using System;
using System.Collections.Generic;

namespace Wardx
{
    public sealed class WardxCore
    {
        readonly Settings _settings;
        readonly ITracer _tracer;
        readonly Dictionary<object, object> _wrappers;
        readonly MetricsRegistry _metrics;
        readonly EventBuffer _events;
        readonly LogBuffer _logs;
        readonly ExperimentResolver _experiments;
        int _seq;
        List<Frame> _pendingFrames = new List<Frame>();
        long _windowStart;
        string _subjectId;

        public readonly InternalMetrics Internal = new InternalMetrics();
        public readonly ConfigStore ConfigStore = new ConfigStore();
        public readonly CoreLogApi Log;

        internal ExperimentResolver ExperimentResolver => _experiments;

        public WardxCore(Settings settings)
        {
            _settings = settings;
            _tracer = settings.Tracer;
            _wrappers = _tracer != null ? new Dictionary<object, object>() : null;
            _metrics = new MetricsRegistry(
                settings.MaxSeriesPerMetric,
                settings.MaxDimensionKeys,
                settings.MaxDimensionValueLength,
                settings.HistogramBuckets,
                () => { Internal.CardinalityDropped += 1; }
            );
            _events = new EventBuffer(settings.MaxBufferedEvents);
            _logs = new LogBuffer(settings.MaxBufferedLogs);
            _experiments = new ExperimentResolver(settings.PrivacySalt, settings.ExperimentStateMaxSubjects, payload =>
            {
                Event("experiment.exposure", payload);
            });
            _windowStart = Clock.UnixMs();
            Log = new CoreLogApi(this);
        }

        public ICounter Counter(string name, IReadOnlyDictionary<string, object> dims = null)
        {
            return Wrap(_metrics.Counter(name, dims), NoopCounter.Instance, name, dims,
                (series, noop) => new TracedCounter(series, _tracer, name, dims, noop));
        }

        public IGauge Gauge(string name, IReadOnlyDictionary<string, object> dims = null)
        {
            return Wrap(_metrics.Gauge(name, dims), NoopGauge.Instance, name, dims,
                (series, noop) => new TracedGauge(series, _tracer, name, dims, noop));
        }

        public IHistogram Histogram(string name, IReadOnlyDictionary<string, object> dims = null, double[] buckets = null)
        {
            return Wrap(_metrics.Histogram(name, dims, buckets), NoopHistogram.Instance, name, dims,
                (series, noop) => new TracedHistogram(series, _tracer, name, dims, noop));
        }

        public TimerToken Timer(string name, IReadOnlyDictionary<string, object> dims = null)
        {
            if (_tracer == null) return _metrics.Timer(name, dims);
            var start = System.Diagnostics.Stopwatch.GetTimestamp();
            return new TimerToken(endDims =>
            {
                var duration = (System.Diagnostics.Stopwatch.GetTimestamp() - start) * 1000.0 / System.Diagnostics.Stopwatch.Frequency;
                Histogram(name, Dimensions.Merge(dims, endDims)).Observe(duration);
            });
        }

        public void Event(string name, IReadOnlyDictionary<string, object> attrs = null)
        {
            var dropped = !_events.Push(name, attrs);
            if (dropped) Internal.EventsDropped += 1;
            TracerEmit.Event(_tracer, new EventRecord { Name = name, Attrs = attrs, Dropped = dropped });
        }

        internal void WriteLog(string level, string message, IReadOnlyDictionary<string, object> attrs)
        {
            var dropped = !_logs.Push(level, message, attrs);
            if (dropped) Internal.LogsDropped += 1;
            TracerEmit.Log(_tracer, new LogRecord { Level = level, Message = message, Attrs = attrs, Dropped = dropped });
        }

        public void Identify(string subjectId)
        {
            if (subjectId != null && subjectId.Length == 0)
            {
                throw new ArgumentException("identify requires a non-empty subjectId");
            }
            _subjectId = subjectId;
        }

        string ResolveSubjectId(string subjectId)
        {
            return subjectId ?? _subjectId;
        }

        public object ConfigGet(string key, object fallback, string subjectId = null)
        {
            if (!ConfigStore.Has(key)) return fallback;
            var remote = ConfigStore.GetRaw(key);
            var resolved = ResolveSubjectId(subjectId);
            if (resolved == null) return remote;
            return _experiments.Resolve(key, remote, resolved, ConfigStore.ExperimentsByKey);
        }

        public T ConfigGet<T>(string key, T fallback, string subjectId = null)
        {
            var value = ConfigGet(key, (object)fallback, subjectId);
            if (value is T typed) return typed;
            if (value == null) return fallback;
            return (T)Convert.ChangeType(value, typeof(T), System.Globalization.CultureInfo.InvariantCulture);
        }

        public void ExperimentGoal(string name, string subjectId = null, object value = null)
        {
            subjectId = ResolveSubjectId(subjectId);
            if (subjectId == null)
            {
                throw new ArgumentException("experiment.goal requires subjectId");
            }
            if (string.IsNullOrEmpty(name))
            {
                throw new ArgumentException("experiment.goal requires a metric name");
            }
            var assignment = _experiments.ExposedAssignmentForGoal(subjectId, name);
            if (assignment == null) return;
            var subject = _experiments.HashSubject(subjectId);
            var payload = new Dictionary<string, object>
            {
                ["metric"] = name,
                ["subject"] = subject,
                ["experiments"] = AssignmentWire(assignment)
            };
            if (value != null) payload["value"] = value;
            Event("experiment.goal", payload);
        }

        public void ApplyConfig(int version, Dictionary<string, object> values, List<ExperimentDefinition> experiments)
        {
            ConfigStore.ApplySnapshot(version, values, experiments);
            _experiments.ApplySnapshot(ConfigStore.Experiments);
            Internal.ConfigVersion = version;
        }

        public FrameBatch SnapshotIfDirty()
        {
            if (!_metrics.IsDirty() && _events.Length == 0 && _logs.Length == 0 && !Internal.HasCounterActivity())
            {
                return null;
            }
            return SnapshotFrame();
        }

        public FrameBatch SnapshotFrame()
        {
            var to = Clock.UnixMs();
            var from = _windowStart;
            _windowStart = to;
            Internal.EventsBuffered = _events.Length;
            Internal.LogsBuffered = _logs.Length;
            var metrics = _metrics.SnapshotAndReset();
            var events = _events.Swap();
            var logs = _logs.Swap();
            var internalSnap = Internal.SnapshotAndReset();
            var frame = FrameBuilder.Build(_seq + 1, from, to, metrics, events, logs, internalSnap);
            var batch = FrameBuilder.SplitToMaxBytes(frame, _settings.MaxFrameBytes);
            _seq = batch.Frames[batch.Frames.Count - 1].Seq;
            _pendingFrames.AddRange(batch.Frames);
            for (int i = 0; i < batch.Frames.Count; i++)
            {
                var physical = batch.Frames[i];
                TracerEmit.Frame(_tracer, new FrameRecord
                {
                    Seq = physical.Seq,
                    From = physical.From,
                    To = physical.To,
                    Counters = physical.Counters.Count,
                    Gauges = physical.Gauges.Count,
                    Histograms = physical.Histograms.Count,
                    Events = physical.Events.Count,
                    Logs = physical.Logs.Count,
                    DroppedLogs = i == 0 ? batch.DroppedLogs : 0,
                    DroppedEvents = i == 0 ? batch.DroppedEvents : 0,
                    DroppedRows = i == 0 ? batch.DroppedRows : 0
                });
            }
            return batch;
        }

        public List<Frame> TakePendingFrames()
        {
            var frames = _pendingFrames;
            _pendingFrames = new List<Frame>();
            return frames;
        }

        T Wrap<T>(T series, T noopSentinel, string name, IReadOnlyDictionary<string, object> dims, Func<T, bool, T> factory)
            where T : class
        {
            if (_tracer == null) return series;
            if (ReferenceEquals(series, noopSentinel)) return factory(series, true);
            if (_wrappers.TryGetValue(series, out var cached)) return (T)cached;
            var wrapped = factory(series, false);
            _wrappers[series] = wrapped;
            return wrapped;
        }

        static List<object> AssignmentWire(ExperimentAssignment assignment)
        {
            return new List<object>
            {
                new Dictionary<string, object>
                {
                    ["experiment"] = assignment.Experiment,
                    ["variant"] = assignment.Variant
                }
            };
        }
    }

    public sealed class CoreLogApi
    {
        readonly WardxCore _core;

        internal CoreLogApi(WardxCore core)
        {
            _core = core;
        }

        public void Debug(string message, IReadOnlyDictionary<string, object> attrs = null)
        {
            _core.WriteLog("debug", message, attrs);
        }

        public void Info(string message, IReadOnlyDictionary<string, object> attrs = null)
        {
            _core.WriteLog("info", message, attrs);
        }

        public void Warn(string message, IReadOnlyDictionary<string, object> attrs = null)
        {
            _core.WriteLog("warn", message, attrs);
        }

        public void Error(string message, IReadOnlyDictionary<string, object> attrs = null)
        {
            _core.WriteLog("error", message, attrs);
        }
    }
}
